### DISCUSSION TITLE
how to change diagnostics starting time in diag_table?

### DISCUSSION THREAD

**Discussion by junwang-noaa** [NOAA-GFDL/FMS#1378] [General]
Hello,  We are setting up an IAU test. It is a restart run from the previous cycle.  Now we want all the diagnostic fields especially those averaged or accumulated fields start to accumulate at a later forecast hour, in other words, not the model starting time. Currently in my diag_table I have:
```
20161003.00Z.C96.64bit.non-mono
2016 10 03 00 0 0
```
2016103000 is the starting time, the model restarts at fh=3, and I want all the averaged fields starts to accumulate at fh=6. Should I change the second line to 2016100306 in my case? e.g.
```
20161003.00Z.C96.64bit.non-mono
2016 10 03 06 0 0
```
I hope with this change the following two files:
```
"ocn%4yr%2mo%2dy%2hr",      6,  "hours", 1, "hours", "time", 6, "hours", "1901 1 1 0 0 0"
"SST%4yr%2mo%2dy",      1,  "days",  1, "days",  "time", 1, "days",  "1901 1 1 0 0 0"
```
will give me 6 hourly diagnostic fields (e.g. 2016100306-2016100312) in ocn files and daily diagnostic fields (2016100306-2016100406) for SST files even though the model starting time is 2016100300 and restarts at fh=03 (2016100303). Thanks.

**Comment by bensonr** [comment]
@thomas-robinson @uramirez8707 - can you guys help out here?

**Reply by uramirez8707** [reply]
I don't think this is working as expected. 
In my test case my simulation start time is `2 1 1 0 0 0`

In my diag_table I have,
```
test_diag_manager
2 1 1 0 0 0

#output files
"ocn_default%4yr%2mo%2dy%2hr",      6,  "hours", 1, "hours", "time", 6, "hours", "2 1 1 3 0 0"
"test_diag_manager_mod", "sst", "sst", "ocn_default%4yr%2mo%2dy%2hr",  "all", .true., "none", 2
```

The `2 1 1 3 0 0` is the [start_time for the file](https://github.com/NOAA-GFDL/FMS/blob/f0ec93f9acb7432bdeed58b879e8cc6599ad8abe/diag_manager/diag_table.F90#L138-L146), so I expected the first file to use data from `2 1 1 3 0 0` to `2 1 1 9 0 0`, but that is not the case:
```
double time_bnds(time, nv) ;
                time_bnds:units = "hours since 0002-01-01 00:00:00" ;
                time_bnds:long_name = "time axis boundaries" 

 time_bnds =
  0, 6,
  6, 12 ;
```

If i change the table to:
```
test_diag_manager
2 1 1 3 0 0

#output files
"ocn_default%4yr%2mo%2dy%2hr",      6,  "hours", 1, "hours", "time", 6, "hours", "2 1 1 3 0 0"
"test_diag_manager_mod", "sst", "sst", "ocn_default%4yr%2mo%2dy%2hr",  "all", .true., "none", 2
```

I get the crash: `FATAL: diag_util_mod::get_date_dif: in variable t2 is less than in variable t1`

If I modify the register_diag_field call from:
```F90
Time = set_date(2,1,1,0,0,0)
id_sst = register_diag_field  ('test_diag_manager_mod', 'sst', (/id_x,id_y,id_z/), Time, 'SST', 'K')
```

to 
``` F90
id_sst = register_diag_field  ('test_diag_manager_mod', 'sst', (/id_x,id_y,id_z/), set_date(2,1,1,3,0,0), 'SST', 'K')
```

I get:
```
double time_bnds(time, nv) ;
                time_bnds:units = "hours since 0002-01-01 00:00:00" ;
                time_bnds:long_name = "time axis boundaries" ;

time_bnds =
  3, 9 ;
```
But i don't think this is a desirable solution.

**Reply by junwang-noaa** [reply]
@uramirez8707 Thanks for the reply. I got the same error with the diag_table change only. So if I understand correctly, in order to change the accumulated diagnostic starting time, we need to
1) specify base time to the diag start time in diag_able
2) change the Time argument in the register_diag_field subroutine call to use the new base time instead of the current model time.

Is this correct? Is it possible that the base time specified in the diat_able can be set as default in the register_diag_field?

**Reply by uramirez8707** [reply]
There is a bug in the diag_manager where the [start time for the file](https://github.com/NOAA-GFDL/FMS/blob/f0ec93f9acb7432bdeed58b879e8cc6599ad8abe/diag_manager/diag_table.F90#L138-L146) is not working as advertised. 

I think with this code update: https://github.com/uramirez8707/FMS/commit/796152a50ae3a3f978ed1b28be2d4f7158485d03

All you need to do
```
"ocn%4yr%2mo%2dy%2hr",      6,  "hours", 1, "hours", "time", 6, "hours", 2016 10 03 06 0 0"
```
With this the file is going to be created from "2016 10 03 06 0 0" regardless of the model start time. 

I think this a better solution.

**Reply by bensonr** [reply]
@uramirez8707 - thanks for the prompt tracking and suggested fix.  I'm assuming our unit tests for this situation are clean with your update.

@junwang-noaa - Please let us know if this works as expected in a functional test of the UFS and we'll work to get this into a planned release of FMS.

**Reply by junwang-noaa** [reply]
@uramirez8707 Thanks for providing the solution. @bensonr I haven't built fms for a while, I will see if I can get some help to build the fms library and  test it in ufs. Thanks.

**Reply by bensonr** [reply]
@junwang-noaa - Jong (@jkbk2004) should be able to assist you in running the UFS with an externally built FMS library.

**Reply by uramirez8707** [reply]
sorry @junwang-noaa my fix is still not ready. Even though the time_bnds are correct [3, 9], it is still using data from hour 0 to hour 9. I need to work on this some more.

**Reply by junwang-noaa** [reply]
@uramirez8707 Thanks for letting me know!

**Reply by uramirez8707** [reply]
The fix is ready to be tested: 
https://github.com/NOAA-GFDL/FMS/pull/1381

**Reply by junwang-noaa** [reply]
@uramirez8707 Thanks, I am working with nceplibs team to build the library. Will let you know what we get (hera is down today).

**Reply by junwang-noaa** [reply]
@uramirez8707 is your branch based on fms 2023.02? I am compiling the UFS weather model with your fms branch, but got error:

```
/scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/20230817/ufs-weather-model/FV3/atmos_cubed_sphere/model/multi_gases.F90(45): error #6580: Name in only-list does not exist or is not accessible.   [OPEN_NAMELIST_FILE]
      use fms_mod,           only: check_nml_error, open_namelist_file, close_file
----------------------------------------------------^
/scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/20230817/ufs-weather-model/FV3/atmos_cubed_sphere/model/multi_gases.F90(45): error #6580: Name in only-list does not exist or is not accessible.   [CLOSE_FILE]
      use fms_mod,           only: check_nml_error, open_namelist_file, close_file
------------------------------------------------------------------------^
compilation aborted for /scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/20230817/ufs-weather-model/FV3/atmos_cubed_sphere/model/multi_gases.F90 (code 1)
make[2]: *** [FV3/atmos_cubed_sphere/CMakeFiles/fv3.dir/model/multi_gases.F90.o] Error 1
```
Then I added "-Duse_deprecated_io" to compile the dycore code, but still got the same error message. Do we need to build when compile FMS? Thanks for your help!

**Comment by bensonr** [comment]
@junwang-noaa - there is a PR into GFDL_atmos_cubed_sphere with updates for the dev/emc branch of GFDL_atmos_cubed_sphere that will update the code to remove the deprecated IO components.  Please see [FV3 PR #272](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/272/files).

If you don't want to cherry pick those commits, you will need to recompile the FMS library with the -Duse_deprecated_io macro.

**Reply by junwang-noaa** [reply]
Thanks, Rusty. Since I am running a coupled test with MOM6, I think I will compile fms with -Duse_deprecated_io macro.

**Comment by junwang-noaa** [comment]
@uramirez8707 I tested your branch in ufs. I think it fixed the starting time issue. In the restart run I have, the starting time 2021032206. the model restarts at fh=03, I want to set average diagnostic field starting time at 2021032212 (skipping accumulating from 2021032209-2021032212) in the ocean history files. I have this in my diag_table:

```
20210322.06Z.C96.64bit.non-mono
2021 03 22 06 0 0

"ocn%4yr%2mo%2dy%2hr",      6,  "hours", 1, "hours", "time", 6, "hours", "2021 3 22 12 0 0"
"SST%4yr%2mo%2dy",      1,  "days",  1, "days",  "time", 1, "days",  "2021 3 22 12 0 0"
```

With your fms branch, I got the following output:
```
ocn_2021_03_22_15.nc
ocn_2021_03_22_21.nc
ocn_2021_03_23_03.nc
ocn_2021_03_23_09.nc
SST_2021_03_23.nc
```
Which looks correct to me. When I looked at each file, e.g. ocn_2021_03_22_15.nc, I see:

```
        double time(time) ;
                time:units = "hours since **2021-03-22 06:00:00**" ;
                time:long_name = "time" ;
                time:axis = "T" ;
                time:calendar_type = "JULIAN" ;
                time:calendar = "julian" ;
                time:bounds = "time_bnds" ;
...
 time = 9 ;

 time_bnds =
  6, 12 ;

```

For the daily average file SST_2021_03_23.nc, I have time_bnds:
```
        double time(time) ;
                time:units = "days since 2021-03-22 06:00:00" ;
...
 time = 0.75 ;
...
 time_bnds =
  0.25, 1.25 ;
```
The results are what I expect. Thanks for making the updates. Are you going to have fms release with the code updates? Thanks

**Reply by uramirez8707** [reply]
@thomas-robinson @rem1776 Can https://github.com/NOAA-GFDL/FMS/pull/1381/files be included in upcoming beta tag and release?

**Comment by junwang-noaa** [comment]
@thomas-robinson @rem1776 May I ask if you have a timeline to include the changes in a tag? Thanks a lot!

**Comment by junwang-noaa** [comment]
@thomas-robinson @rem1776 Is it possible that you can create a beta release for us with this updates? We need this for GFSv17 implementation. The issue becomes a blocker to create correct mom6 output files. Thanks for your help!

**Reply by rem1776** [reply]
@junwang-noaa Sorry for the delay! We are doing initial beta testing for the next release now, if we end up doing another beta tag we will include these changes.

Otherwise if we do go forward with the release, we can likely merge this in right after and create a new tag to test with.

Either way we should be able to get these changes in a tag in a week or so.

**Reply by junwang-noaa** [reply]
Thanks a lot! We will plan accordingly.

**Comment by thomas-robinson** [comment]
@junwang-noaa are you running with default real\*4?  The 2023.03 release will update FMS to support both real\*4 and real\*8 which *may* cause some answer changes when compiling/running the components with r4.

**Comment by junwang-noaa** [comment]
Thanks for letting me know. In our coupled tests, we are running with real*4 in fv3 dycore and real*8 in MOM6. If you have a beta testing branch, I can test it in UFS.

**Comment by thomas-robinson** [comment]
@junwang-noaa we are going to patch FMS 2023.02 and create FMS 2023.02.01 for the least impact to your work.  @uramirez8707 and @rem1776 are working on it now, so hopefully it will be ready today or Monday.  This should not impact your real4 fv3 runs like moving to 2023.03 might.

**Reply by junwang-noaa** [reply]
That is great! Thank you very much!

**Reply by rem1776** [reply]
The patch has been [released](https://github.com/NOAA-GFDL/FMS/releases/tag/2023.02.01), to use it you would just pull the tag `2023.02.01`. Let us know if you have any issues!

**Reply by junwang-noaa** [reply]
Thank you! We are going to install and test it, will let you know if there is any issues.
